File: 2020-04/4-09/22括号生成/22.py
#  -*- coding: utf-8 -*-

#  Editor      : Pycharm
#  File        : 22.py
#  Author      : ZhenyuMa
#  Created     : 2020/4/9 8:51 下午
#  Description : 一道dfs + 剪枝


from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        self.res = []
        self.helper(n, n, '')
        return self.res

    def helper(self, left_residual, right_residual, cur_res):
        if left_residual == 0 and right_residual == 0:
            self.res.append(cur_res)
            return

        if left_residual > right_residual:
            return

        if left_residual > 0:
            self.helper(left_residual - 1, right_residual, cur_res + '(')
        if right_residual > 0:
            self.helper(left_residual, right_residual - 1, cur_res + ')')

File: 2020-04/4-29/22括号生成/22.cpp


class Solution {
private:
    vector<string> res;
public:
    vector<string> generateParenthesis(int n) {
        string cur_res = "";
        helper(n, n, cur_res);
        return res;
    }

    void helper(int left_residual, int right_residual, string cur_res){
        if(left_residual == 0 && right_residual == 0){
            res.push_back(cur_res);
            return;
        }

        // 剪枝，因为已经不可能是有效括号了。
        if(left_residual > right_residual)
            return;

        if(left_residual > 0)
            helper(left_residual - 1, right_residual, cur_res + '(');
        if(right_residual > 0)
            helper(left_residual, right_residual - 1, cur_res + ')');
    }
};